x-pack/filebeat/tests/integration/managerV2_test.go

```diff
@@ -344,12 +344,162 @@ func TestFailedOutputReportsUnhealthy(t *testing.T) {
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "filestream-input",
+			Type: "filestream",
+			Name: "filestream-input-healty",
+			Streams: []*proto.Stream{
+				{
+					Id: "filestream-input-id",
+					Source: integration.RequireNewStruct(t, map[string]interface{}{
+						"id":      "filestream-stream-input-id",
+						"enabled": true,
+						"type":    "filestream",
+						"paths":   logFilePath,
+					}),
+				},
+			},
+		},
+	}
+
+	filestreamInputStarting := proto.UnitExpected{
+		Id:             "input-unit-2",
+		Type:           proto.UnitType_INPUT,
+		ConfigStateIdx: 1,
+		State:          proto.State_STARTING,
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "filestream-input",
+			Type: "filestream",
+			Name: "filestream-input-starting",
+			Streams: []*proto.Stream{
+				{
+					Id: "filestream-input-id",
+					Source: integration.RequireNewStruct(t, map[string]interface{}{
+						"id":      "filestream-stream-input-id",
+						"enabled": true,
+						"type":    "filestream",
+						"paths":   logFilePath,
+					}),
+				},
+			},
+		},
+	}
+
+	healthyOutput := proto.UnitExpected{
+		Id:             "output-unit",
+		Type:           proto.UnitType_OUTPUT,
+		ConfigStateIdx: 1,
+		State:          proto.State_HEALTHY,
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "default",
+			Type: "elasticsearch",
+			Name: "elasticsearch",
+			Source: integration.RequireNewStruct(t,
+				map[string]interface{}{
+					"type":     "elasticsearch",
+					"hosts":    []interface{}{"http://localhost:9200"},
+					"username": "admin",
+					"password": "testing",
+					"protocol": "http",
+					"enabled":  true,
+				}),
+		},
 	}
-	return str
+
+	brokenOutput := proto.UnitExpected{
+		Id:             "output-unit-borken",
+		Type:           proto.UnitType_OUTPUT,
+		ConfigStateIdx: 1,
+		State:          proto.State_FAILED,
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "default",
+			Type: "logstash",
+			Name: "logstash",
+			Source: integration.RequireNewStruct(t,
+				map[string]interface{}{
+					"type":    "logstash",
+					"invalid": "configuration",
+				}),
+		},
+	}
+
+	// Those are the 'states' Filebeat will go through.
+	// After each state is reached the mockServer will
+	// send the next.
+	protoUnits := [][]*proto.UnitExpected{
+		{
+			&healthyOutput,
+			&filestreamInputHealthy,
+		},
+		{
+			&brokenOutput,
+			&filestreamInputStarting,
+		},
+		{
+			&healthyOutput,
+			&filestreamInputHealthy,
+		},
+		{}, // An empty one makes the Beat exit
+	}
+
+	// We use `testDone` to notify `require.Eventually` at the end of the test
+	// that the desired state has been reached.
+	testDone := atomic.MakeBool(false)
+	// The test is successful when we reach the last element of `protoUnits`
+	onObserved := func(observed *proto.CheckinObserved, protoUnitsIdx int) {
+		if protoUnitsIdx == len(protoUnits)-1 {
```

Comment:
Shouldn't you verify that the output is actually healthy? Or are you relying on the fact that there was a kind of deadlock before?

If you are relying on the test deadlocking, that doesn't actually protect us against introducing a new bug later that only breaks the output reload. Imagine somebody accidentally deletes or moves the lines you added to unconditionally save the output configuration for example.

Reply:
The server returned by `integration.NewMockServer` already ensures all units reach the state set on their `proto.UnitExpected.State`. If you look at the definitions of every unit, they all have the final state set, e.g: https://github.com/elastic/beats/blob/e3367eddcb6f03c3d7c38e9735999be1a9ca1cb9/x-pack/filebeat/tests/integration/managerV2_test.go#L410-L430. The server will only send the next array of units when all units reach their desired state.

This `onObserved` function allows the test to perform extra operations on every checking.